pandas/tests/frame/methods/test_astype.py

```diff
@@ -726,3 +727,32 @@ def test_astype_categorical_to_string_missing(self):
         cat = df.astype("category")
         result = cat.astype(str)
         tm.assert_frame_equal(result, expected)
+
+
+class IntegerArrayNoCopy(pd.core.arrays.IntegerArray):
+    # GH 42501
+
+    @classmethod
+    def _from_sequence(cls, scalars, *, dtype=None, copy=False):
+        values, mask = coerce_to_array(scalars, dtype=dtype, copy=copy)
+        return IntegerArrayNoCopy(values, mask)
+
+    def copy(self):
+        assert False
+
+
+class Int16DtypeNoCopy(pd.Int16Dtype):
+    # GH 42501
+
+    @classmethod
+    def construct_array_type(cls):
+        return IntegerArrayNoCopy
+
+
+def test_frame_astype_no_copy():
+    # GH 42501
+    df = DataFrame({"a": [1, 4, None, 5], "b": [6, 7, 8, 9]}, dtype=object)
+    result = df.astype({"a": Int16DtypeNoCopy()}, copy=False)
+
+    tm.assert_frame_equal(df, result, check_dtype=False)
+    tm.assert_numpy_array_equal(df.b.values, result.b.values, check_same="same")
```

Comment:
this check doesnt' tell us anything. you can check that these *share* memory if you wish (would be useful).

Reply:
Oh, checking that the memory is shared is certainly my intention with this test.  Per the `assert_numpy_array_equal` doc string:

```
    check_same : None|'copy'|'same', default None
        Ensure left and right refer/do not refer to the same memory area.
```

This test does fail on 1.3.0 and pass with this change. Is there a better way to do it or I'm misunderstanding something?

Reply:
oh i c, forgot about arg (we usually simply use `np.shares_memory` (but this is ok, maybe add a comment that is what you are testing).